petastorm/reader_impl/worker_loop.py

```diff
@@ -0,0 +1,170 @@
+        # rowgroup_spec is currently a ParquetDatasetPiece, but will likely be augmented with petastorm specific info
+        # in the future
+        try:
+            rowgroup_spec = next(epochs_iterator)
+        except StopIteration:
+            rowgroup_spec = None
+
+        # Continue iterating until we have some data in one of the queues connecting loader/shuffling-queue/decoder or
+        # we still have some rowgroup_specs to process
+        while rowgroup_spec is not None or in_loading_futures or in_decoding or shuffling_queue.size > 0:
+            # An owner can signal a stop before we are actually finished
+            if stop_event.is_set():
+                break
+
+            # 1. Read more rowgroup_specs from our epochs generator and schedule 'loader' to load these rowgroups
+            # 2. Readout what was already loaded and enqueue loaded rows into a shuffling queue
+            # 3. Read from the shuffling queue and submit decoding tasks to decoding executor
+            # 4. Read completed decoding futures and put the results into the final output queue
+
+            # 1. Read more rowgroup_specs from our epochs generator and schedule 'loader' to load these rowgroups
+            # ---------------------------------------------------------------------------------------------------
+
+            # Need to keep a cap (_MAX_IN_LOADING_QUEUE_SIZE) on the queue so we don't get unbounded growth with
+            # infinite epochs.
+            if rowgroup_spec is not None and len(in_loading_futures) < _MAX_IN_LOADING_QUEUE_SIZE:
+                in_loading_futures.append(loading_pool.submit(loader.load, rowgroup_spec))
+                stats['rowgroups_scheduled_for_loading'] += 1
+
+                try:
+                    rowgroup_spec = next(epochs_iterator)
+                except StopIteration:
+                    rowgroup_spec = None
+
+            # 2. Readout what was already loaded and enqueue loaded rows into a shuffling queue
+            # ---------------------------------------------------------------------------------
+
+            # Futures that have completed will be placed first in the returned list
+            loaded_futures = as_completed(in_loading_futures, timeout=0)
+
+            # Push all loaded_futures and partially decoded data into shuffling queue
+            while shuffling_queue.can_add():
+                try:
+                    loaded_future = next(loaded_futures)
+                    loaded_future_result = loaded_future.result()
+                    stats['rows_loaded'] += len(loaded_future_result)
+                    shuffling_queue.add_many(loaded_future_result)
+
+                    # Done processing, remove from the queue. 'remove' on the list is ok since we cap the list
+                    # to a small size with _MAX_IN_LOADING_QUEUE_SIZE
+                    in_loading_futures.remove(loaded_future)
+                except (TimeoutError, StopIteration):
```

Comment:
Do we need to catch something else here in case loading the data threw a different error?

Reply:
If an unexpected exception is thrown, I think we need to fail the whole thing.